File: frontend/src/stores/localStateStore.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { LoadLocalState, SaveLocalState } from '../../bindings/github.com/ys-ll/uniterm/app'

// Shape of the backend's LocalState store (bindings are untyped JS, so the
// type is declared here and must stay in sync with backend/store).
interface LocalState {
  sidebarVisible: boolean
  aiSidebarVisible: boolean
  // Whether the bottom bar (second panel area) is shown. Optional: a
  // local_state.json written before the bar existed has no such key, and a
  // missing value means "shown" (see DEFAULT below).
  bottomBarVisible?: boolean
  collapsedGroupIds: string[]
  collapsedQuickCommandGroupIds: string[]
  windowX: number
  windowY: number
  windowWidth: number
  windowHeight: number
  windowMaximised: boolean
  backgroundEnabled: boolean
  backgroundImage: string
  backgroundOpacity: number
  backgroundBlur: number
  backgroundFit: string
  systemTitleBar: boolean
  externalEditor: string
  // SFTP file-list columns the user hid via the header context menu. Columns
  // absent from the list stay visible.
  sftpHiddenColumns: string[]
}

const DEFAULT: LocalState = {
  sidebarVisible: true,
  aiSidebarVisible: true,
  bottomBarVisible: true,
  collapsedGroupIds: [],
  collapsedQuickCommandGroupIds: [],
  windowX: 0,
  windowY: 0,
  windowWidth: 0,
  windowHeight: 0,
  windowMaximised: false,
  backgroundEnabled: false,
  backgroundImage: '',
  backgroundOpacity: 60,
  backgroundBlur: 3,
  backgroundFit: 'cover',
  systemTitleBar: false,
  externalEditor: '',
  sftpHiddenColumns: [],
} as LocalState

export const useLocalStateStore = defineStore('localState', () => {
  const state = ref<LocalState>({ ...DEFAULT })
  const loaded = ref(false)
  let initPromise: Promise<void> | null = null

  async function init() {
    if (loaded.value) return
    if (!initPromise) {
      initPromise = (async () => {
        try {
          const s = await LoadLocalState()
          state.value = { ...DEFAULT, ...s } as LocalState
        } catch {
          // keep defaults
        } finally {
          loaded.value = true
        }
      })()
    }
    return initPromise
  }

  async function update(patch: Partial<LocalState>) {
    state.value = { ...state.value, ...patch } as LocalState
    try {
      await SaveLocalState(state.value as LocalState)
    } catch {
      // ignore save errors
    }
  }

  return { state, loaded, init, update }
})

File: backend/store/local_state_store.go
package store

import (
	"encoding/json"
	"os"
	"path/filepath"
)

const localStateFileName = "local_state.json"

type LocalState struct {
	SidebarVisible   bool `json:"sidebarVisible"`
	AISidebarVisible bool `json:"aiSidebarVisible"`
	// BottomBarVisible remembers whether the bottom bar (the second, resizable
	// panel area) is shown. Pointer + omitempty so local_state.json written by
	// older builds still loads; nil means "shown", which the frontend falls
	// back to.
	BottomBarVisible  *bool    `json:"bottomBarVisible,omitempty"`
	CollapsedGroupIds []string `json:"collapsedGroupIds"`
	// Collapsed quick-command group ids (plus "__ungrouped__"). Local-only
	// UI state, never synced; groups absent from the list stay expanded.
	CollapsedQuickCommandGroupIds []string `json:"collapsedQuickCommandGroupIds"`
	WindowX                       int      `json:"windowX"`
	WindowY                       int      `json:"windowY"`
	WindowWidth                   int      `json:"windowWidth"`
	WindowHeight                  int      `json:"windowHeight"`
	WindowMaximised               bool     `json:"windowMaximised"`
	// Background image — local-only appearance, never synced.
	BackgroundEnabled bool   `json:"backgroundEnabled"`
	BackgroundImage   string `json:"backgroundImage"`
	BackgroundOpacity int    `json:"backgroundOpacity"`
	BackgroundBlur    int    `json:"backgroundBlur"`
	BackgroundFit     string `json:"backgroundFit"`
	// SystemTitleBar switches the window to the OS native frame instead of
	// the built-in one. Read at startup only — changing it needs a restart.
	SystemTitleBar bool `json:"systemTitleBar"`
	// ExternalEditor command used to open remote files in an external editor
	// (SFTP "edit externally"). Local-only preference, never synced.
	ExternalEditor string `json:"externalEditor"`
	// SftpHiddenColumns lists SFTP file-list columns hidden via the header
	// context menu (type/modTime/size/permission/owner/group). Local-only UI
	// state, never synced; columns absent from the list stay visible.
	SftpHiddenColumns []string `json:"sftpHiddenColumns,omitempty"`
}

type LocalStateStore struct {
	configDir string
}

func NewLocalStateStore(configDir string) *LocalStateStore {
	return &LocalStateStore{configDir: configDir}
}

func (s *LocalStateStore) filePath() string {
	return filepath.Join(s.configDir, localStateFileName)
}

func (s *LocalStateStore) Save(state LocalState) error {
	bytes, err := json.MarshalIndent(state, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(s.filePath(), bytes, 0600)
}

func defaultLocalState() LocalState {
	return LocalState{
		SidebarVisible:                true,
		AISidebarVisible:              true,
		CollapsedQuickCommandGroupIds: []string{},
		BackgroundOpacity:             60,
		BackgroundBlur:                3,
		BackgroundFit:                 "cover",
	}
}

func (s *LocalStateStore) Load() (LocalState, error) {
	bytes, err := os.ReadFile(s.filePath())
	if err != nil {
		if os.IsNotExist(err) {
			return defaultLocalState(), nil
		}
		return LocalState{}, err
	}
	// Start from defaults so fields missing in older configs keep their
	// intended values instead of Go zero values (e.g. opacity/blur = 0).
	state := defaultLocalState()
	if err := json.Unmarshal(bytes, &state); err != nil {
		return defaultLocalState(), nil
	}
	return state, nil
}
